### PR TITLE
feat: add get-charm-paths action (KF-4012)

### PR DESCRIPTION
**Pull request by ca-scribner**
Adds an action for finding all the charm paths in a repo.  Useful for building a gh CI matrix, as shown in readme.md

### CODE REVIEW

File: actions/get-charm-paths/src/get_charm_paths.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#!/bin/env python
+import argparse
+from pathlib import Path
+
+
+OUTPUT_VARIABLE_NAME = "charm_paths"
+
+
+def find_charms_in_dir(base_dir: str, charms_subdir="charms"):
```
Comment (ColmBhandal) [inline actions/get-charm-paths/src/get_charm_paths.py:9]: Does it make sense to also default `base_dir` to the root dir?

**Review by ColmBhandal — DISMISSED** [review]
I left some comments inline

File: actions/get-charm-paths/action.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    description: 'JSON string of relative paths to charms in the repository'
+    value: ${{ steps.get-charm-paths.outputs.charm_paths }}
+runs:
+  using: 'composite'
+  steps:
+    - uses: 'actions/checkout@v2'
+    - name: 'Get paths for all charms in repo'
+      id: get-charm-paths
+      shell: bash
+      run: python ${{ github.action_path }}/src/get_charm_paths.py ./
```
[the remark this answers is not in the document]
Comment (ca-scribner) [reply actions/get-charm-paths/action.yml:14]: I think(?) you have to assume.  You could install python, but then you'll leave python as a side effect on their machine

practically speaking, I think all standard gh runners have python.  Could be an issue for a custom runner though

File: actions/get-charm-paths/README.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    runs-on: ubuntu-latest
+    outputs:
+      charm_paths: ${{ steps.get-charm-paths.outputs.charm-paths }}
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Get paths for all charms in this repo
+        id: get-charm-paths
+        uses: canonical/kubeflow-ci/actions/get-charm-paths@master
```
Comment (ColmBhandal) [inline actions/get-charm-paths/README.md:27]: Replace `master` -> `main`

File: actions/get-charm-paths/src/get_charm_paths.py (hunk 1 above)
Comment (ca-scribner) [reply actions/get-charm-paths/src/get_charm_paths.py:9]: yeah so `base_dir: str = "./"`?  lgtm

File: actions/get-charm-paths/action.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ca-scribner) [reply actions/get-charm-paths/action.yml:14]: yep lgtm on everything.  Accept arguments for base_dir and charms_subdir (with defaults of `./` and `charms`), then always provide both to the underlying script?  that sounds great

File: actions/get-charm-paths/action.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+name: 'Get Charm Paths'
+description: 'Emits a list of relative paths to charms in the repository'
+outputs:
+  charm-paths:
+    description: 'JSON string of relative paths to charms in the repository'
+    value: ${{ steps.get-charm-paths.outputs.charm_paths }}
+runs:
+  using: 'composite'
+  steps:
+    - uses: 'actions/checkout@v2'
```
[the remark this answers is not in the document]
Comment (ColmBhandal) [reply actions/get-charm-paths/action.yml]: Done

File: actions/get-charm-paths/src/get_charm_paths.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+def stringify_paths(paths):
+	"""Converts list of Paths to list of strings, appending a trailing slash just in case"""
+	return [str(path) + "/" for path in paths]
+
+
+def emit_to_github_action(name, data):
+	"""Emits data to STDOUT as an output named name in a format that GitHub Actions consumes"""
+	data = str(data)
+	print(f"Found {name}: {data}")
+	print(f"::set-output name={name}::{data}")
```
[the remark this answers is not in the document]
Comment (ColmBhandal) [reply actions/get-charm-paths/src/get_charm_paths.py]: Done

**Review by ColmBhandal — COMMENTED** [review]
Removing request for chagnes

File: actions/get-charm-paths/src/get_charm_paths.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    * If not, then if the base directory itself is a charm directory, that is returned.
+    * Otherwise, returns []
+
+	Directories returned are alphabetically sorted.
+    """
+    base_dir = Path(base_dir)
+    charms_dir = base_dir / charms_subdir
+    
+    # Look for both .yaml and .yml files in subdirectories using a single glob pattern
+    metadata_files = list(charms_dir.glob("*/metadata.y*ml"))
```
Comment (misohu) [inline actions/get-charm-paths/src/get_charm_paths.py:26]: I think this regex matches much more than "yaml" and "yml" ... e.g yeml or yiml or yoml. But its minor think I still think we can merge.

**Review by misohu — DISMISSED** [review]

**Review by ColmBhandal — APPROVED** [review]

**Pull request merged**